libparsec/crates/client/src/workspace/store/resolve_path.rs

```diff
@@ -727,69 +795,273 @@ fn cache_only_retrieve_path_from_id(
+    loop {
+        // Most of the time we should have each entry in the path already in the cache,
+        // so we want to lock the cache once and only release it in the unlikely case
+        // we need to fetch from the local storage or server.
+        let cache_only_outcome = {
+            let mut cache = store.current_view_cache.lock().expect("Mutex is poisoned");
+            cache_only_retrieve_path_from_id(
+                &mut cache,
+                entry_id,
+                false,
+                last_not_found_during_populate,
+            )
+        };
+        match cache_only_outcome {
+            CacheOnlyPathRetrievalOutcome::Done {
+                entry,
+                maybe_update_lock_guard,
+            } => {
+                debug_assert!(maybe_update_lock_guard.is_none());
+                return Ok(entry);
             }
+
             // We got a cache miss
-            CacheOnlyRetrievalOutcome::NeedPopulateCache(cache_miss_entry_id) => {
-                populate_cache_from_local_storage_or_server(store, cache_miss_entry_id)
-                    .await
-                    .map_err(|err| match err {
-                        PopulateCacheFromLocalStorageOrServerError::Offline => {
-                            ResolvePathError::Offline
-                        }
-                        PopulateCacheFromLocalStorageOrServerError::Stopped => {
-                            ResolvePathError::Stopped
-                        }
-                        PopulateCacheFromLocalStorageOrServerError::EntryNotFound => {
-                            ResolvePathError::EntryNotFound
-                        }
-                        PopulateCacheFromLocalStorageOrServerError::NoRealmAccess => {
-                            ResolvePathError::NoRealmAccess
-                        }
-                        PopulateCacheFromLocalStorageOrServerError::InvalidKeysBundle(err) => {
-                            ResolvePathError::InvalidKeysBundle(err)
-                        }
-                        PopulateCacheFromLocalStorageOrServerError::InvalidCertificate(err) => {
-                            ResolvePathError::InvalidCertificate(err)
-                        }
-                        PopulateCacheFromLocalStorageOrServerError::InvalidManifest(err) => {
-                            ResolvePathError::InvalidManifest(err)
-                        }
-                        PopulateCacheFromLocalStorageOrServerError::Internal(err) => {
-                            err.context("cannot fetch manifest").into()
-                        }
-                    })?;
+            CacheOnlyPathRetrievalOutcome::NeedPopulateCache(cache_miss_entry_id) => {
+                match populate_cache_from_local_storage_or_server(store, cache_miss_entry_id).await
+                {
+                    // Happy case :)
+                    Ok(_) => (),
+                    // The entry seems to just not exist, we cannot just
```

Comment:
we cannot just...?

Reply:
don't try to argue, WE CANNOT JUST !!!! :trollface:

Reply:
done.